examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+
+        loss = outputs if isinstance(outputs, torch.Tensor) else outputs["loss"]
+
+        self.fabric.call("on_before_backward", loss)
+        self.fabric.backward(loss)
+        self.fabric.call("on_after_backward")
+        # TODO: reroute configure_gradient_clipping in LM to fabric.gradient clipping
+        # self.fabric.call('configure_gradient_clipping')
+
+        # avoid gradients in stored/accumulated values -> prevents potential OOM
+        self._current_train_return = apply_to_collection(outputs, dtype=torch.Tensor, function=lambda x: x.detach())
+
+        return loss
+
+    def step_scheduler(
+        self,
+        model: LightningModule,
+        scheduler_cfg: Optional[Mapping[str, Union[LRScheduler, bool, str, int]]],
+        level: Literal["step", "epoch"],
+        current_value: int,
+    ):
+        # no scheduler
+        if scheduler_cfg is None:
+            return
+
+        # wrong interval (step vs. epoch)
+        if scheduler_cfg["interval"] != level:
+            return
+
+        # right interval, but wrong step wrt frequency
+        if current_value % cast(int, scheduler_cfg["frequency"]) != 0:
+            return
+
+        # assemble potential monitored values
+        possible_monitor_vals = {None: None}
+        if isinstance(self._current_train_return, torch.Tensor):
+            possible_monitor_vals.update("train_loss", self._current_train_return)
+        elif isinstance(self._current_train_return, Mapping):
+            possible_monitor_vals.update({"train_" + k: v for k, v in self._current_train_return.items()})
+
+        if isinstance(self._current_val_return, torch.Tensor):
+            possible_monitor_vals.update("val_loss", self._current_val_return)
+        elif isinstance(self._current_val_return, Mapping):
+            possible_monitor_vals.update({"val_" + k: v for k, v in self._current_val_return.items()})
+
+        try:
+            monitor = possible_monitor_vals[cast(Optional[str], scheduler_cfg["monitor"])]
+        except KeyError as e:
+            possible_keys = list(possible_monitor_vals.keys())
+            raise KeyError(
+                f"monitor {scheduler_cfg['monitor']} is invalid. Possible values are {possible_keys}."
+            ) from e
+
+        # rely on model hook for actual step
+        model.lr_scheduler_step(scheduler_cfg["scheduler"], monitor)
+
+    @property
+    def should_validate(self) -> bool:
+        return self.current_epoch % self.validation_frequency == 0
+
+    def pbar_wrapper(self, iterable, total, **kwargs):
+        if self.fabric.is_global_zero:
+            return tqdm(iterable, total=total, **kwargs)
+        return iterable
+
+    def load(self, state, path):
+        if state is None:
+            state = {}
+
+        remainder = self.fabric.load(path, state)
+        self.global_step = remainder.pop("global_step")
+        self.current_epoch = remainder.pop("current_epoch")
```

Comment:
I think if you included these in the initial state, you could always save everything. And when you load here, the items would be reloaded in place.

Reply:
answered above :)